src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -2991,6 +2996,15 @@ void InterpreterSelectQuery::ignoreWithTotals()
     getSelectQuery().group_by_with_totals = false;
 }
 
+bool InterpreterSelectQuery::forceSelectFinalOnSelectQuery(ASTSelectQuery & query)
+{
+    // query.tables() is required because not all queries have tables in it, it could be a function.
+    auto is_force_select_final_setting_on = context->getSettingsRef().force_select_final;
+    auto is_final_supported = storage && storage->supportsFinal() && !storage->isRemote() && query.tables();
+    auto is_query_already_final = query.final();
+
+    return is_force_select_final_setting_on && !is_query_already_final && is_final_supported;
```

Comment:
This is not correct because there can be several tables, one of them supporting `final`, and another not supporting it.

Reply:
This is also already supported, I have added a stateless test. Please check: https://github.com/ClickHouse/ClickHouse/pull/40945/files#diff-637c4a076d097fc4303b8af39daeef42a2b2b0de9bd7394257ef6621194c9361R50-R68